archive_upload/handlers/dsmc_handlers.py

```diff
@@ -734,21 +735,27 @@ def exclude_content(tarinfo):
         log.info("Removing files from {} that were added to {}".format(
             path_to_archive_root, tarball_path))
 
-        # Remove files that we added to the tarball.
-        with tarfile.open(tarball_path) as tar:
-            for member in tar.getmembers():
-                try:
-                    filepath = os.path.normpath(
-                        os.path.join(path_to_archive_root, archive, member.name))
-
-                    if os.path.isfile(filepath):
-                        os.remove(filepath)
-                    elif member.name != "." and os.path.isdir(filepath):
-                        self._rm_empty_dirs(filepath)
-
-                except OSError, e:
-                    msg = "Error when creating archive tarball. Could not remove file {}: {}".format(filepath, e)
-                    raise ArchiveException(reason=msg, status_code=500)
+        # Remove files that were added to the tarball
+
+        # get a list of files and folders duplicated between the tarball and the archive
+        paths_to_remove = FileUtils.paths_duplicated_in_tarball(tarball_path, path_to_archive)
+
+        # remove the files first
+        for file_to_remove in filter(os.path.isfile, paths_to_remove):
+            try:
+                os.remove(file_to_remove)
+            except OSError as e:
+                raise ArchiveException(
+                    status_code=500,
+                    reason="Error when creating archive tarball. Could not remove file {}: {}".format(
+                        file_to_remove, e))
+
+        # then the directories
+        for dir_to_remove_if_empty in filter(os.path.isdir, paths_to_remove):
+            try:
+                os.rmdir(dir_to_remove_if_empty)
+            except OSError as e:
+                log.debug("directory {} not removed, probably because it is not empty".format(dir_to_remove_if_empty))
```

Comment:
I think maybe this should have `warning` as it's log level, since it is actually an error that we might be interested in seeing even if the service is not in debug mode?

Reply:
Also, when we cannot remove a file we raise an Exception, which will cause a 500 from the service. But when we cannot remove a dir we only log it. This seems a bit inconsistent to me.

Reply:
Isn't it a possible scenario that the directory contains files that are not in the tarball?
If you have this file structure
```
+-- foo
|   +-- bar.txt
|   +-- baz.txt
+-- fizz
|   +-- buzz
```
and the tar only has file `bar.txt` from `foo`, but not `baz.txt`, then `foo` would show up in the list returned by `paths_duplicated_in_tarball()`

Reply:
yes, exactly. not being able to remove a directory is expected, hence the debug level, whereas failing to remove a file is not and an exception is thrown

Reply:
Ok. Then I get it. Let me know when you want me to merge.